File: 0079-word-search/0079-word-search.py
class Solution(object):
    def exist(self, board, word):
        m = len(board)
        n = len(board[0])

        def dfs(i, r, c):
            if i == len(word):
                return True
            if not (0 <= r < m) or not (0 <= c < n):
                return False
            if board[r][c] == '' or board[r][c] != word[i]:
                return False

            temp = board[r][c]
            board[r][c] = '' 

            res = (
                dfs(i + 1, r + 1, c) or
                dfs(i + 1, r - 1, c) or
                dfs(i + 1, r, c + 1) or
                dfs(i + 1, r, c - 1)
            )

            board[r][c] = temp  # restore original character
            return res

        for r in range(m):
            for c in range(n):
                if dfs(0, r, c):
                    return True

        return False

File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool dfs(
        int index,
        vector<vector<char>>& board,
        string& word,
        int row,
        int col
    ) {
        if (index == word.size()) {
            return true;
        }

        if (
            row < 0 || col < 0 ||
            row >= board.size() || col >= board[0].size() ||
            board[row][col] != word[index]
        ) {
            return false;
        }

        char temp = board[row][col];
        board[row][col] = '&';

        bool result =
            dfs(index + 1, board, word, row, col + 1) ||
            dfs(index + 1, board, word, row + 1, col) ||
            dfs(index + 1, board, word, row - 1, col) ||
            dfs(index + 1, board, word, row, col - 1);

        board[row][col] = temp;

        return result;
    }

    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[i].size(); j++) {
                if (dfs(0, board, word, i, j)) {
                    return true;
                }
            }
        }

        return false;
    }
};